### DISCUSSION TITLE
Strongswan FIPS on RHEL

### DISCUSSION THREAD

**Discussion by Pankaj-Deshpande-Com** [strongswan/strongswan#1716] [Q&A] [answered]
**Tobias and other Strongswan Experts,** 

I have a few queries about Strongswan FIPS on RHEL 8 OS.  We found Strongswan FIPS  and Ubuntu is documented here: https://ubuntu.com/security/fips#modules But I didn't find similar document for Strongswan on  RHEL. 

Our product is RHEL based and we use Strongswan as our inflight encryption stack as an important building block. Our customers are asking for FIPS compliance so we are planning to enable FIPS mode in our RHEL kernel. 

I have following queries:
1. Once we enable FIPS mode in our RHEL 8 kernel, would our Strongswan based IPSec stack work? 
2. If yes, would Strongswan/RHEL be FIPS compliant (as in using only FIPS-Validated modules)?  
3. If not, Can you please help provide changes we may have to do in Strongswan so as to get it working on rhel in FIPS mode? 
4. Is there any existing Strongswan FIPS on RHEL effort  that you would like us to be aware of?

I know there are other native IKE daemons on RHEL that are FIPS compliant on RHEL but we are keen on Strongswan and our customers are pushing for FIPS. Can you help in anyways ? 
 
Regards,
Pankaj

**Comment by tobiasbrunner** [comment, marked as answer]
> 1. Once we enable FIPS mode in our RHEL 8 kernel, would our Strongswan based IPSec stack work?

Since strongSwan runs in userland, this has no effect on it directly (i.e. strongSwan does not change its behavior). Since the kernel will disable several algorithms for use with IPsec, you have to make sure not to negotiate such algorithms for ESP (otherwise, the installation of the CHILD_SAs will fail).

> 2. If yes, would Strongswan/RHEL be FIPS compliant (as in using only FIPS-Validated modules)?

It can be. For instance, if all crypto for IKE is handled by a FIPS-certified crypto library (e.g. OpenSSL). In particular, the IKEv2 key derivation (possible to be handled by an external library since 5.9.6, older versions required separate FIPS-validation of that step, see e.g. the Ubuntu certificate). Make sure you disable the default crypto plugins (e.g. _sha1_, _sha2_, _aes_ etc.) for that.

**Reply by Pankaj-Deshpande-Com** [reply]
Thanks Tobias. 

**Few more queries:** 

TB wrote> Make sure you disable the default crypto plugins (e.g. sha1, sha2, aes etc.) for that.
**`Yes we have all default crypto plugins disabled.**

TB wrote>For instance, if all crypto for IKE is handled by a FIPS-certified crypto library (e.g. OpenSSL).
**`Would OpenSSL sufficient if we are planning to use standard, minimal plugins ? Is there any guidance to get it running with minimalist support/plugins/config/suites ?  In particular, would like to understand what other compliant crypto libraries we may need to integrate with beyond openssl, if at all**?

Our product:
- Runs on AWS RHEL 8.6 EC2 VM. 
- For IKE SA, We use "aes256-sha256-ecp256" only  (as both sides of tunnels are controlled by us.)
- For Child SA, We are AES-NI (no offload) based. We use "aes256gcm16-prfsha256-ecp256" only (again, as both sides of tunnels are controlled by us).

As our customers are asking for FIPS in AWS, Azure, **So we are looking for shortest path by configuring strongswan (minimalistically) that can support our above mentioned config** .

**Reply by tobiasbrunner** [reply]
You can see the [algorithms implemented by the plugins](https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html) in the docs. The ones you intend to use are fully supported by the _openssl_ plugin (only IKE is relevant as IPsec is handled by the kernel - by the way, you don't need to specify a PRF in the ESP proposal `aes256gcm16-ecp256` is enough). So you don't need any additional crypto libraries.

**Reply by Pankaj-Deshpande-Com** [reply]
Thanks for your quick reply. This helps !
